alerts/config.go

```diff
@@ -5,116 +5,179 @@ package alerts
-	NoCommunication *WithDelay             `json:"noCommunication,omitempty" bson:"noCommunication,omitempty"`
+	FollowedID string `json:"followedId" bson:"followedId"`
+
+	Alerts `bson:",inline,omitempty"`
+}
+
+// Alerts models a user's desired alerts.
+type Alerts struct {
+	UrgentLow       *UrgentLowAlert       `json:"urgentLow,omitempty" bson:"urgentLow,omitempty"`
+	Low             *LowAlert             `json:"low,omitempty" bson:"low,omitempty"`
+	High            *HighAlert            `json:"high,omitempty" bson:"high,omitempty"`
+	NotLooping      *NotLoopingAlert      `json:"notLooping,omitempty" bson:"notLooping,omitempty"`
+	NoCommunication *NoCommunicationAlert `json:"noCommunication,omitempty" bson:"noCommunication,omitempty"`
 }
 
 func (c Config) Validate(validator structure.Validator) {
 	validator.String("UserID", &c.UserID).Using(user.IDValidator)
 	validator.String("FollowedID", &c.FollowedID).Using(user.IDValidator)
-	if c.Low != nil {
-		c.Low.Validate(validator)
+	c.Alerts.Validate(validator)
+}
+
+func (i Alerts) Validate(validator structure.Validator) {
+	if i.Low != nil {
+		i.Low.Validate(validator)
 	}
-	if c.UrgentLow != nil {
-		c.UrgentLow.Validate(validator)
+	if i.UrgentLow != nil {
+		i.UrgentLow.Validate(validator)
 	}
-	if c.High != nil {
-		c.High.Validate(validator)
+	if i.High != nil {
+		i.High.Validate(validator)
 	}
-	if c.NotLooping != nil {
-		c.NotLooping.Validate(validator)
+	if i.NotLooping != nil {
+		i.NotLooping.Validate(validator)
 	}
-	if c.NoCommunication != nil {
-		c.NoCommunication.Validate(validator)
+	if i.NoCommunication != nil {
+		i.NoCommunication.Validate(validator)
 	}
 }
 
 // Base describes the minimum specifics of a desired alert.
 type Base struct {
 	// Enabled controls whether notifications should be sent for this alert.
-	Enabled bool
+	Enabled bool `json:"enabled"`
 	// Repeat is measured in minutes.
-	Repeat DurationMinutes `json:"repeat"`
+	//
+	// A value of 0 (the default) disables repeat notifications.
+	Repeat DurationMinutes `json:"repeat,omitempty"`
 }
 
 func (b Base) Validate(validator structure.Validator) {
 	validator.Bool("enabled", &b.Enabled)
 	dur := b.Repeat.Duration()
-	validator.Duration("repeat", &dur).GreaterThan(0 * time.Minute)
+	validator.Duration("repeat", &dur).Using(validateRepeat)
 }
 
-// DelayMixin adds a configurable delay.
-type DelayMixin struct {
-	// Delay is measured in minutes.
-	Delay DurationMinutes `json:"delay,omitempty"`
-}
+const (
+	// RepeatMin is the minimum duration for a repeat setting (if not 0).
+	RepeatMin = 15 * time.Minute
+	// RepeatMax is the maximum duration for a repeat setting.
+	RepeatMax = 4 * time.Hour
+)
 
-func (d DelayMixin) Validate(validator structure.Validator) {
-	dur := d.Delay.Duration()
-	validator.Duration("delay", &dur).GreaterThan(0 * time.Minute)
+func validateRepeat(value time.Duration, errorReporter structure.ErrorReporter) {
+	if value == 0 {
```

Comment:
Sure, I can do that. I generally don't find it useful to make constants for Zero values in Go, as 99% of the time it's either a default or an invalid value. I'm just trained to read Go code that way, and I find it removes a bit of cruft. Since that's not a shared experience, I can define a const.

Reply:
Yea, that makes sense where using 0 (or the zero-value) is obvious and it is not being used as a "special value".

In this case I don't think it is obvious what 0 actually means, at least to me. Since 0 is the zero-value for a `time.Duration`, if the field is missing in JSON, does it mean "use a system-defined default value to repeat after", "never repeat", or something else? 

For me to know the answer I have to scan backwards through the code and find the comment `// A value of 0 (the default) disables repeat notifications.` However, if the constant was something like `RepeatDurationNeverRepeat` or `RepeatDurationDisabled` then the code becomes self-documenting and a separate manual comment becomes less critical for understanding. (Oh, yes, before I forget, I suck at naming things, but you get the idea. ;) )

Alternatively, you could be completely clear by using a separate boolean to indicate if repeating was enabled/disabled. That makes it even clearer, but at the expense of a bit of complexity.

Reply:
> In this case I don't think it is obvious what 0 actually means, at least to me. Since 0 is the zero-value for a `time.Duration`, if the field is missing in JSON, does it mean "use a system-defined default value to repeat after", "never repeat", or something else?

That's interesting to me. When I read this code, 0 is very clearly "don't repeat". I can't even imagine what else it could mean. But I have no problem with adding a const. Brains are strange things. :)